compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2769,6 +2805,146 @@ BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType tar
         }
     }
 
+    BType getTypeIntersectedWithMapType(BMapType mapType, BType targetType, SymbolEnv env) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BType intersectionConstraintType = getTypeIntersection(mapType.constraint,
+                    ((BMapType) targetType).constraint, env);
+            if (intersectionConstraintType == symTable.semanticError) {
+                return symTable.semanticError;
+            }
+            return new BMapType(TypeTags.MAP, intersectionConstraintType, null);
+        }
+        return getTypeIntersectedWithRecordType((BRecordType) targetType, mapType, env);
```

Comment:
Aren't we repeating some checks by doing this? Wouldn't it be better to extract out the map-record intersection logic to a separate method and directly call that here and in `getTypeIntersectedWithRecordType`?

Reply:
Seems like `getTypeIntersectedWithRecordType` is only called from this method. If so we can change the `targetType` param of `getTypeIntersectedWithRecordType` to be of a `BMapType` and remove the tag checks.

Reply:
Refactored